runtime/lua/vim/fs.lua

```diff
@@ -40,15 +42,34 @@ function M.dirname(file)
   if file == nil then
     return nil
   end
-  return vim.fn.fnamemodify(file, ':h')
+  vim.validate({ file = { file, 's' } })
+  if file:match('^%w:[\\/]?$') and iswin then
```

Comment:
Probably is only a minor consideration, but maybe put `iswin` in the condition first, so that it fails quickly when not on Windows (otherwise, it will run `string.match()` *every* time, even if `iswin` is false).